eclair-core/src/test/scala/fr/acinq/eclair/blockchain/bitcoind/ZmqWatcherSpec.scala

```diff
@@ -146,8 +155,102 @@ class ZmqWatcherSpec extends TestKitBaseClass with AnyFunSuiteLike with Bitcoind
+    listener.expectMsgAllOf(
+      WatchEventSpentBasic(BITCOIN_FUNDING_SPENT),
+      WatchEventSpent(BITCOIN_FUNDING_SPENT, tx2)
+    )
+    system.stop(watcher)
+  }
+
+  test("watch for unknown spent transactions") {
+    val probe = TestProbe()
+    val blockCount = new AtomicLong()
+    val wallet = new BitcoinCoreWallet(bitcoinrpcclient)
+    val client = new ExtendedBitcoinClient(bitcoinrpcclient)
+    val watcher = system.actorOf(ZmqWatcher.props(blockCount, client))
+
+    // create a chain of transactions that we don't broadcast yet
+    val (_, priv) = getNewAddress(bitcoincli)
+    val tx1 = {
+      wallet.fundTransaction(Transaction(2, Nil, TxOut(150000 sat, Script.pay2wpkh(priv.publicKey)) :: Nil, 0), lockUnspents = true, FeeratePerKw(250 sat)).pipeTo(probe.ref)
+      val funded = probe.expectMsgType[FundTransactionResponse].tx
+      wallet.signTransaction(funded).pipeTo(probe.ref)
+      probe.expectMsgType[SignTransactionResponse].tx
+    }
+    val outputIndex = tx1.txOut.indexWhere(_.publicKeyScript == Script.write(Script.pay2wpkh(priv.publicKey)))
+    val tx2 = createSpendP2WPKH(tx1, priv, priv.publicKey, 10000 sat, 1, 0)
+
+    // setup watches before we publish transactions
+    probe.send(watcher, WatchSpent(probe.ref, tx1, outputIndex, BITCOIN_FUNDING_SPENT))
+    probe.send(watcher, WatchConfirmed(probe.ref, tx1, 3, BITCOIN_FUNDING_SPENT))
+    client.publishTransaction(tx1).pipeTo(probe.ref)
+    probe.expectMsg(tx1.txid)
+    generateBlocks(bitcoincli, 1)
+    probe.expectNoMsg(1 second)
+    client.publishTransaction(tx2).pipeTo(probe.ref)
+    probe.expectMsgAllOf(tx2.txid, WatchEventSpent(BITCOIN_FUNDING_SPENT, tx2))
+    probe.expectNoMsg(1 second)
+    generateBlocks(bitcoincli, 1)
+    probe.expectMsg(WatchEventSpent(BITCOIN_FUNDING_SPENT, tx2)) // tx2 is confirmed which triggers WatchEventSpent again
+    generateBlocks(bitcoincli, 1)
+    assert(probe.expectMsgType[WatchEventConfirmed].tx === tx1) // tx1 now has 3 confirmations
+    system.stop(watcher)
+  }
+
+  test("publish transactions with relative and absolute delays") {
+    val probe = TestProbe()
+    val blockCount = new AtomicLong()
+    val wallet = new BitcoinCoreWallet(bitcoinrpcclient)
+    val client = new ExtendedBitcoinClient(bitcoinrpcclient)
+    val watcher = system.actorOf(ZmqWatcher.props(blockCount, client))
+    awaitCond(blockCount.get > 0)
+    val (_, priv) = getNewAddress(bitcoincli)
+
+    // tx1 has an absolute delay but no relative delay
+    val tx1 = {
+      wallet.fundTransaction(Transaction(2, Nil, TxOut(150000 sat, Script.pay2wpkh(priv.publicKey)) :: Nil, blockCount.get + 5), lockUnspents = true, FeeratePerKw(250 sat)).pipeTo(probe.ref)
+      val funded = probe.expectMsgType[FundTransactionResponse].tx
+      wallet.signTransaction(funded).pipeTo(probe.ref)
+      probe.expectMsgType[SignTransactionResponse].tx
+    }
+    probe.send(watcher, PublishAsap(tx1))
+    generateBlocks(bitcoincli, 5)
```

Comment:
Regarding my comment about potential off-by-one errors, how about:
```suggestion
    probe.send(watcher, PublishAsap(tx1))
    generateBlocks(bitcoincli, 4)
    probe.expectNoMsg()
    generateBlocks(bitcoincli, 1)
```

Maybe too paranoid?

Reply:
Nope, that makes sense!
Even though if we end up broadcasting too soon, `bitcoind` will reject the tx and many tests will fail.
If we end up broadcasting too late, then we're wasting a 10 minutes delay but at least we're safe :)

Reply:
There are probably several places where this pattern could be used.

Reply:
Done in https://github.com/ACINQ/eclair/pull/1500/commits/fc4e6b0c72d3d743124f8615141e8f858eb9d03c, this is really the only test that benefited explicitly from this.